### PR TITLE
[WIP] Python3 support

### LINKED ISSUE #2807: Cannot depend on virtual package and non-virtual package
If A depends on B and C, and B depends on a virtual package while C depends on a specific implementation of that package, Spack will fail during concretization. This example was discovered with `py-yt`. `py-yt` depends on `py-matplotlib` which depends on `py-pillow`. `py-yt` also depends on `py-scikit-image`, which depends on `pil`. Even though `py-pillow` is the default implementation of `pil`, Spack crashes with the following message:
```
==> Error: Cannot depend on 'py-pillow^binutils@2.26:^bison^bzip2@1.0.6%gcc@6.1.0 arch=linux-centos6-x86_64^flex^gettext@0.19:^help2man^m4^ncurses@6.0%gcc@6.1.0 arch=linux-centos6-x86_64^openssl@1.0.2j%gcc@6.1.0 arch=linux-centos6-x86_64^py-setuptools@25.2.0%gcc@6.1.0 arch=linux-centos6-x86_64^python@2.7.13%gcc@6.1.0+tk~ucs4 arch=linux-centos6-x86_64^readline@6.3%gcc@6.1.0 arch=linux-centos6-x86_64^sqlite@3.8.5%gcc@6.1.0 arch=linux-centos6-x86_64^tcl^tk@8.3:^zlib@1.2.10%gcc@6.1.0+pic arch=linux-centos6-x86_64' twice
```

### PR DESCRIPTION
**Pull request by tgamblin**
Fixes #2807.

I am cleaning up the concretizer, and part of that is cleaning up `package_prefs` and the use of `cmp`/`__cmp__` on `Specs`.  It turns out this is also something we need for Python 3.

So I went ahead and did all the easy stuff to get spack to work in Python 3.  I'm still working on the refactoring parts.  

Currently, Spack will run if invoked with `python3`, but it'll fail when it hits the first `sorted(..., cmp=...)` call.  I think there should be relatively few major issues getting Python3 to work after that is resolved, and we'll have easier code to port for the new concretizer.

- [x] remove version guard in `bin/spack`
- [x] update YAML version in externals
- [x] port `ordereddict_backport`
- [x] vendor [`six`](https://pypi.python.org/pypi/six) to aid with porting
- [x] Fix StringIO, print, except as, octals, izip calls, xrange, etc.
- [x] use `six.string_types` instead of `basestring`
- [x] use `six.iteritems`
- [x] Remove assumptions about `ascii` in the code base
- [x] Fix packages that have moved (`urllib`, etc.)
- [x] Fix UTF-8 issues with `proc.communicate` in `Executable`
- [x] Remove use of `cmp` in the code
- [x] Try to use `str` nearly everywhere in the spack code; avoid `unicode` objects in 2.x, but handle utf-8 gracefully in Python 3.
- [x] Refactor and simplify handling of deptypes in the concretizer
- [x] Fix #2807: ordering issues with merging providers & vdeps in concretization.
- [x] Fix minimum version checking (add `pyqver3`)
- [x] Clean up a number of spec tests to NOT depend on DB ordering
- [x] Add Python3 to Travis testing

### CODE REVIEW

**Comment by tgamblin** [comment]
Ok tests are passing for 2.6-2.7 and 3.2-3.6.

* @skosukhin: can you take a look at dd53381106c3eb025c93084dcbc3c1d6ee6083be, particularly the parts to do with `spack_json`?  This relates to your #2524.
* @adamjstewart: can you verify that this fixes #2807?

File: lib/spack/spack/concretize.py [hunk 1]
```diff
@@ -34,86 +34,78 @@
...
-        usable.sort(cmp=cmp_externals)
-        return usable
+        # Use a sort key to order the results
+        return sorted(usable, key=lambda spec: (
+            not (spec.external or spec.external_module),  # prefer externals
+            pref_key(spec),                               # respect prefs
+            spec.name,                                    # group by name
+            reverse_order(spec.versions),                 # latest version
+            spec                                          # natural order
+        ))
```
Comment (alalazo) [inline lib/spack/spack/concretize.py]: I would give +10 for the readability improvement here!

File: var/spack/repos/builtin/packages/elfutils/package.py [hunk 2]
```diff
@@ -40,7 +40,7 @@ class Elfutils(AutotoolsPackage):
     list_depth = 2
 
     version('0.168','52adfa40758d0d39e5d5c57689bf38d6')
-    version('0.163','77ce87f259987d2e54e4d87b86cbee41')
+    version('0.163','77ce87f259987d2e54e4d87b86cbee41', preferred=True)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/elfutils/package.py]: Is this change intentional?

**Comment by skosukhin** [comment]
@tgamblin fine with me. Maybe it's better to edit the comments as well, i.e. the word "byteified" should be changed to "strify" since you've changed the name of the function.

File: lib/spack/docs/conf.py [hunk 3]
```diff
@@ -110,13 +114,13 @@
...
         if not line.startswith('    :noindex:'):
-            print '    :noindex: %s' % ' '.join(spack.__all__)
+            print('    :noindex: %s' % ' '.join(spack.__all__))
         handling_spack = False
 
     if line.startswith('.. automodule::'):
         handling_spack = (line == '.. automodule:: spack\n')
 
-    print line,
+    sys.stdout.write(line)
```
Comment (tgamblin) [inline lib/spack/docs/conf.py]: There's no way to do print-with-comma with the new print function.  Using write() omits the newline.

File: var/spack/repos/builtin/packages/elfutils/package.py (hunk 2 above)
Comment (tgamblin) [inline var/spack/repos/builtin/packages/elfutils/package.py]: It should probably be a separate PR. Livdwarf doesn't build properly with the latest elfutils due to header issues.  This makes the build succeed in the tests.

**Comment by tgamblin** [comment]
@mplegendre: can you look over the package prefs and see whether I am missing anything?
